lang/src/org/partiql/lang/eval/Thunk.kt

```diff
@@ -37,167 +38,186 @@ typealias ThunkEnv = (Environment) -> ExprValue
 typealias ThunkEnvValue<T> = (Environment, T) -> ExprValue
 
 /**
- * Creates a [ThunkEnv], which handles exceptions by wrapping them into an [EvaluationException] which uses
- * [handleException] to handle exceptions appropriately.
+ * A thunk excepting a configurable exception handler (([Throwable], [SourceLocationMeta]?) -> [Nothing])
+ *
+ * The default exception handler wraps any [Throwable] exception and throws [EvaluationException]
  *
- * Literal lambdas passed to this function as [t] are inlined into the body of the function being returned, which
- * reduces the need to create additional call contexts.  The lambdas passed as [t] may not contain non-local returns
- * (`crossinline`).
+ * This name was chosen because it is a thunk which accepts exception handler options as its arguments.
  */
-internal inline fun thunkEnv(metas: MetaContainer, crossinline t: ThunkEnv): ThunkEnv {
-    val sourceLocationMeta = metas.find(SourceLocationMeta.TAG) as? SourceLocationMeta
-
-    return { env ->
-        handleException(sourceLocationMeta) {
-            t(env)
-        }
-    }
+class ThunkOptions(
+        val handleException: (Throwable, SourceLocationMeta?) -> Nothing = defaultExceptionHandler // <-- the default has the same behavior we currently have.
+)
+
+val defaultExceptionHandler: (Throwable, SourceLocationMeta?) -> Nothing = { e, sourceLocation ->
+    val message = e.message ?: "<NO MESSAGE>"
+    throw EvaluationException(
+        "Internal error, $message",
+        errorContext = errorContextFrom(sourceLocation),
+        cause = e,
+        internal = true)
 }
 
 /**
- * Similar to [thunkEnv], but creates a [ThunkEnvValue] instead.
+ * Contains all the thunk methods.
```

Comment:
perhaps: "Provides methods for constructing new thunks according to the specified [ThunkOptions]."?